scripts/proto/lbann_proto.py

```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
-    by child modules to customize adding parents.
-
-    """
-
-    def __init__(self, name, data_layout):
-        self.name = name
-        self.data_layout = data_layout
-        self.parents = []
-        self.children = []
+    def __init__(self, name='', inputs=[], outputs=[],
+                 weights=[], data_layout=''):
+        Layer.num_layers += 1
+        self.name = name if name else 'layer{0}'.format(Layer.num_layers-1)
+        self.inputs = []
+        self.outputs = []
         self.weights = []
+        self.data_layout = data_layout
 
-    def __call__(self, parent):
-        """Make parent a parent of this module."""
-        self.add_parent(parent)
-        return self
-
-    def add_parent(self, parent):
-        """Make parent a parent of this module."""
-        self.parents.append(parent)
-        parent.children.append(self)
-
-    def proto(self, model):
-        """Add this module to the protobuf model."""
-        raise NotImplementedError('proto not implemented')
-
-    def add_weights(self, initializer, name_suffix=None):
-        """Create a new weights object and add it to this module."""
-        if name_suffix is None:
-            name_suffix = '_weights{0}'.format(len(self.weights) + 1)
-        w = Weights(self.name + name_suffix, initializer)
+        # Initialize inputs, outputs, and weights
+        for l in (inputs
+                  if isinstance(inputs, collections.Iterable)
+                  else [inputs]):
+            self.add_input(l)
+        for l in (outputs
+                  if isinstance(outputs, collections.Iterable)
+                  else [outputs]):
+            self.add_output(output)
+        for w in (weights
+                  if isinstance(weights, collections.Iterable)
+                  else [weights]):
+            self.add_weights(w)
+
+    def export_proto(self):
+        """Construct protobuf message for the layer."""
+        proto = lbann_pb2.Layer()
+        proto.parents = ' '.join([l.name for l in self.inputs])
+        proto.children = ' '.join([l.name for l in self.outputs])
+        proto.weights = ' '.join([w.name for w in self.weights])
+        proto.name = self.name
+        proto.data_layout = self.data_layout
+        return proto
+
+    def add_input(self, input):
+        """This layer will receive an input tensor from 'input'."""
+        self.inputs.append(input)
+        input.outputs.append(self)
+
+    def add_output(self, output):
+        """"This layer will send an output tensor to 'output'."""
+        self.outputs.append(output)
+        output.inputs.append(self)
+
+    def add_weights(self, w):
         self.weights.append(w)
-        return w
-
-    def get_weights(self):
-        """Return the weights associated with this module."""
-        return self.weights
 
-    def _add_layer(self, model):
-        """Add a layer entry to model (for use by proto).
+def _create_layer_subclass(type_name, layer_field_name):
```

Comment:
This definitely needed improved documentation, otherwise I'd look at it in a month and it would be black magic. Thanks! 😄